### PR TITLE
New role for hosted cluster on baremetal

### PR DESCRIPTION
**Pull request by Sandeepyadav93**
##### SUMMARY

New role to create hosted guest clusters with BM workers.

##### ISSUE TYPE

- New or Enhanced Feature

##### Tests
- [] TestBos2 - https://www.distributed-ci.io/jobs/4a673808-3418-4b77-82a8-0bae96247f2b/jobStates

Test-Hints: no-check

### CODE REVIEW

File: roles/hostedbm/tasks/post-deploy-step-on-guest-cluster.yml [hunk 1]
```diff
@@ -0,0 +1,115 @@
+---
+- name: "deploy-operators : Install Metallb Operator on the Guest Cluster"
+  ansible.builtin.include_role:
+    name: olm_operator
```
Comment (betoredhat) [inline roles/hostedbm/tasks/post-deploy-step-on-guest-cluster.yml:4]: To use the role from the collection please use the FQCN: redhatci.ocp.olm_operator

File: roles/hostedbm/tasks/post-deploy-step-on-guest-cluster.yml [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        KUBECONFIG: "{{ hostedbm_out_dir }}/{{ hostedbm_cluster_name }}-kubeadmin-kubeconfig"
+  vars:
+    channel: stable
+    install_approval: Automatic
+    operator: metallb-operator
+    namespace: metallb-system
+    source: redhat-operators
+    source_ns: openshift-marketplace
+
+- name: "Set the MetalLB Instance on the Guest Cluster"
```
Comment (betoredhat) [inline roles/hostedbm/tasks/post-deploy-step-on-guest-cluster.yml:16]: The readme says that the metallb must be installed and configured. Can this be optional or what will be the behavior if the MLB is already configured?

There is already a role in the collection to configure metallb with support for L2 and BGP mode, do you think that can be reused here? https://github.com/redhatci/ansible-collection-redhatci-ocp/tree/main/roles/metallb_setup

File: roles/hostedbm/tasks/create-cluster.yml [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    dest: "{{ hostedbm_working_dir }}/hosted_cluster_rendered.yaml"
+
+# replace nodeport with loadbalancer
+- name: Run Bash script to update Kubeapiservice
+  command: |
+    sed -i '/^\s*- service: APIServer$/,/^\s*type: NodePort$/{s/^\(\s*type:\s*\)NodePort$/\1LoadBalancer/; /^\s*nodePort:/, /^\s*address:/d}' {{ hostedbm_working_dir }}/hosted_cluster_rendered.yaml
+
+- name: Create Guest Cluster
+  shell: >-
+    oc --kubeconfig={{ hostedbm_kubeconfig_file }} apply -f {{ hostedbm_working_dir }}/hosted_cluster_rendered.yaml
```
Comment (betoredhat) [inline roles/hostedbm/tasks/create-cluster.yml:35]: Can you use the k8s module instead?

File: roles/hostedbm/tasks/validations.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  assert:
+    that:
+      - hostedbm_kubeconfig_file is defined
+    fail_msg: "The variable hostedbm_kubeconfig_file is not defined. Please define it to proceed."
+
+- name: Make sure working directory doesn't exits from previous deployment
+  file:
+    path: "{{ hostedbm_working_dir }}"
+    state: absent
+
```
Comment (betoredhat) [inline roles/hostedbm/tasks/validations.yml:30]: May be worth validating that the defined storage class is available

File: roles/hostedbm/README.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+------------
+
+* Access to an OpenShift hosting cluster
+* Properly configured kubeconfig file for the hosting cluster access
+* Existing CatalogSource.
+* The OpenShift hosting cluster should have a way to configure Persistent Volumes (PVs), which could be through the ODF (OpenShift Data Foundation), LVM Storage Operator, or any other suitable method.
+* ACM with MCE installed on the hosting cluster
+* Metallb operator installed on the hosting cluster
+* Metallb configured on the hosting cluster - MetalLB, L2Advertisement, and IPAddressPool configured for range of IP to be used as `APIServer` for hosted clusters.
+* AgentServiceConfig and Provisioning configured on the hosting cluster
```
Comment (betoredhat) [inline roles/hostedbm/README.md:16]: Can the role validate that ACM, MCE, ASG, and provisioning are in place?

**Comment by betoredhat** [comment]
@Sandeepyadav93 thanks for your contribution!

Some issues with the linting need to be checked, please see the checks tab. Before start testing the role in our connected environment the references to other roles that are already part of the collection need to be fully qualified. 

Regards

File: roles/hostedbm/tasks/create-cluster.yml (hunk 3 above)
Comment (Sandeepyadav93) [reply roles/hostedbm/tasks/create-cluster.yml]: done

File: roles/hostedbm/tasks/post-deploy-step-on-guest-cluster.yml (hunk 1 above)
Comment (Sandeepyadav93) [reply roles/hostedbm/tasks/post-deploy-step-on-guest-cluster.yml]: done

File: roles/hostedbm/tasks/post-deploy-step-on-guest-cluster.yml (hunk 2 above)
Comment (Sandeepyadav93) [reply roles/hostedbm/tasks/post-deploy-step-on-guest-cluster.yml:16]: @betoredhat 


>> The readme says that the metallb must be installed and configured. Can this be optional or what will be the behavior if the MLB is already configured?

We need to install the MetalLB operator on both the hosting and guest clusters. Before executing this role, we assume that you have already configured MetalLB on the hosting cluster (we reuse the `metallb_setup` role for installing on the hosting cluster). In the `post-deploy-step-on-guest-cluster.yml` task, we install MetalLB on the guest cluster.

>> There is already a role in the collection to configure MetalLB with support for L2 and BGP modes. Do you think that role can be reused here? [metallb_setup role](https://github.com/redhatci/ansible-collection-redhatci-ocp/tree/main/roles/metallb_setup)

For the guest cluster, we need to create an `IPAddressPool` with the following additional spec, which this role currently doesn’t support:

```
        protocol: layer2
        autoAssign: false
```

Comment (Sandeepyadav93) [reply roles/hostedbm/tasks/post-deploy-step-on-guest-cluster.yml:16]: We can move this to usage of `metallb_setup role` once that role support above parameters for IPAddresspool

File: roles/hostedbm/README.md (hunk 5 above)
Comment (Sandeepyadav93) [reply roles/hostedbm/README.md:16]: Thank you for your suggestion.

Actually, we internally reuse this code in our automation. Before calling this hostedbm role, we invoke another role that sets up ACM, configures LVMS storage, and installs and configures MetalLB. We validate the presence of ACM, ASG, and provisioning objects within our ACM installation role.

In my humble opinion, validating everything mentioned in the requirements for this role would make it lengthy, time-consuming, and would also result in duplicated validation across different roles.

I assume you have roles for ACM installation where you perform these object validations.

But if that's the standard in this repo, I don't have strong objections to adding those validations here. Please let me know what you think.

File: roles/hostedbm/tasks/post-deploy-step-on-guest-cluster.yml (hunk 2 above)
Comment (betoredhat) [reply roles/hostedbm/tasks/post-deploy-step-on-guest-cluster.yml:16]: @Sandeepyadav93 Understood. Is the installation of MetalLB required on the guest cluster? If not, could we make all the post tasks optional using a flag? Ideally, the default action of the role should focus on the guest cluster installation.

Yes, we need to work on the metallb_setup role to support those specs.

File: roles/hostedbm/README.md (hunk 5 above)
Comment (betoredhat) [reply roles/hostedbm/README.md:16]: That works, yes. The other roles have the required validations.

File: roles/hostedbm/tasks/post-deploy-step-on-guest-cluster.yml (hunk 2 above)
Comment (Sandeepyadav93) [reply roles/hostedbm/tasks/post-deploy-step-on-guest-cluster.yml:16]: @betoredhat MetalLB is required on the guest cluster, as some operators such as console and ingress will not enter a healthy state without it.

Comment (betoredhat) [reply roles/hostedbm/tasks/post-deploy-step-on-guest-cluster.yml:16]: ok, that sounds good. We haven't had that requirement in our tests with the kubevirt provisioner. But should be good. Let me work on a change to be able to test this role in our lab. Cheers

File: roles/hostedbm/tasks/create-cluster.yml [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+- name: Save output to YAML file
+  ansible.builtin.copy:
+    content: "{{ _hostedbm_hcp_output.stdout }}"
+    dest: "{{ hostedbm_working_dir }}/hosted_cluster_rendered.yaml"
+    mode: "0644"
+
+# replace nodeport with loadbalancer
+- name: Run Bash script to update Kubeapiservice
+  ansible.builtin.command: >
```
Comment (reihl) [inline roles/hostedbm/tasks/create-cluster.yml:31]: Why not using: `ansible.builtin.replace` ?

File: roles/hostedbm/tasks/validations.yml [hunk 7]
```diff
@@ -0,0 +1,41 @@
+---
+- name: Ensure hostedbm_cluster_name is defined
```
Comment (reihl) [inline roles/hostedbm/tasks/validations.yml:2]: From hear till line 24 you can combine it to one task with loop and dictionary
{condition evaluation [true/false],failed messages}

Comment (Sandeepyadav93) [reply roles/hostedbm/tasks/validations.yml]: thank you for the great suggestion, done.

File: roles/hostedbm/tasks/create-cluster.yml (hunk 6 above)
Comment (Sandeepyadav93) [reply roles/hostedbm/tasks/create-cluster.yml:31]: @reihl 

I tried various combination with replace and lineinfile, But I couldn't manage to do the below with ansible replace:-

~~~
sed -i '/^\s*- service: APIServer$/,/^\s*type: NodePort$/{s/^\(\s*type:\s*\)NodePort$/\1LoadBalancer/; /^\s*nodePort:/, /^\s*address:/d}' myorig.yaml
~~~

cat myorig.yaml
~~~
---
apiVersion: hypershift.openshift.io/v1beta1
kind: HostedCluster
metadata:
  name: testing
  namespace: clusters
spec:
  services:
  - service: APIServer
    servicePublishingStrategy:
      nodePort:
        address: 192.168.98.73
      type: NodePort
  - service: OAuthServer
    servicePublishingStrategy:
      nodePort:
        address: 192.168.98.73
      type: NodePort
  - service: OIDC
    servicePublishingStrategy:
      nodePort:
        address: 192.168.98.73
      type: None
  - service: Konnectivity
    servicePublishingStrategy:
      nodePort:
        address: 192.168.98.73
      type: NodePort
  - service: Ignition
    servicePublishingStrategy:
      nodePort:
        address: 192.168.98.73
      type: NodePort
  - service: OVNSbDb
    servicePublishingStrategy:
      nodePort:
        address: 192.168.98.73
      type: NodePort
  sshKey:
    name: cnfdr36-ssh-key
~~~


After modification:-
~~~
---
apiVersion: hypershift.openshift.io/v1beta1
kind: HostedCluster
metadata:
  name: testing
  namespace: clusters
spec:
  services:
  - service: APIServer
    servicePublishingStrategy:
      type: LoadBalancer
  - service: OAuthServer
    servicePublishingStrategy:
      nodePort:
        address: 192.168.98.73
      type: NodePort
  - service: OIDC
    servicePublishingStrategy:
      nodePort:
        address: 192.168.98.73
      type: None
  - service: Konnectivity
    servicePublishingStrategy:
      nodePort:
        address: 192.168.98.73
      type: NodePort
  - service: Ignition
    servicePublishingStrategy:
      nodePort:
        address: 192.168.98.73
      type: NodePort
  - service: OVNSbDb
    servicePublishingStrategy:
      nodePort:
        address: 192.168.98.73
      type: NodePort
  sshKey:
    name: cnfdr36-ssh-key
~~~


If you can help with code how to achieve above via ansible, I am happy to replace the code with native ansible module.

File: roles/hostedbm/tasks/validations.yml [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    that:
+      - "{{ item }} is defined"
+    fail_msg: "The variable {{ item }} is not defined. Please define it to proceed."
+  loop:
+    - hostedbm_cluster_name
+    - hostedbm_cluster_base_domain
+    - hostedbm_guest_ingress_ip
+    - hostedbm_kubeconfig_file
+
+- name: Make sure working directory doesn't exits from previous deployment
```
Comment (reihl) [inline roles/hostedbm/tasks/validations.yml:13]: You can collapse this task as well by using:
```
state: "{{ item }}"
loop:
  - absent
  - directory
```

File: roles/hostedbm/tasks/create-cluster.yml (hunk 6 above)
Comment (Sandeepyadav93) [reply roles/hostedbm/tasks/create-cluster.yml]: Thank you so much!

File: roles/hostedbm/tasks/validations.yml (hunk 8 above)
Comment (Sandeepyadav93) [reply roles/hostedbm/tasks/validations.yml]: done

**Comment by betoredhat** [comment]
recheck

**Comment by betoredhat** [comment]
Hello @Sandeepyadav93 we have validated the role in our CI. In order to merge this, please rebase your PR to pass the Sanith checks as now we are using 2.17. Thanks!

**Review by betoredhat — APPROVED** [review]
lgtm, thanks for your contribution!

**Comment by Sandeepyadav93** [comment]
@betoredhat thank you so much for doing the tests, I have rebased my PR.

**Comment by fredericlepied** [comment]
@Sandeepyadav93 can you sign your commit?

**Comment by Sandeepyadav93** [comment]
@fredericlepied thanks, done.

**Pull request merged**